### ISSUE TITLE
enhancement: high-DPI plots

### ISSUE THREAD

**Issue by curtis999** [rstudio/rstudio#3896]
### System details

    RStudio Edition : Server
    RStudio Version : 1.2.1114
    OS Version      : Ubuntu 18.04.1
    R Version       : 3.5.1

### Steps to reproduce the problem

On the client side all UI icons seem to be high-resolution except for the plot outputs (eg. inline within the R notebook or in the plot panel).

It would be really nice if the Rstudio-server could display retina-quality plots like the desktop version of Rstudio.

(redirected from [https://community.rstudio.com/t/hidpi-plots-within-rstudio-server/18113/2](https://community.rstudio.com/t/hidpi-plots-within-rstudio-server/18113/2))

**Comment by ronblum** [comment]
@curtis999 Thank you for porting over this enhancement request from the Community site!

**Comment by hrbrmstr** [comment]
Count me in as another requester for this. I'm likely going to be moving to 100% RStudio Server and not having high-DPI plots in the viewer will slow down a few interactive daily workflows. #ty

**Comment by kevinushey** [comment]
Now that high DPI displays are becoming more ubiquitous even outside the Apple world, I think we'll need to take this work on soon.

**Comment by gusberinger** [comment]
@curtis999 How were you able to get high DPI plots in the plot panel in RStudio Desktop?

**Comment by curtis999** [comment]
> @curtis999 How were you able to get high DPI plots in the plot panel in RStudio Desktop?

I think high DPI is on by default for me in RStudio Desktop (macOS).

**Comment by valerie-rstudio** [comment]
### Verify Failed

    RStudio Edition : Desktop
    RStudio Version : 1.5.111
    OS Version      : Windows 10
    R Version       : 4.0.3

### Failure Points
- Plot outputs on Windows still look like they're in low resolution.

<hr>

### Verified Fixed

    RStudio Edition : Desktop & Server
    RStudio Version : 1.5.111 & 1.5.109-2
    OS Version      : ubuntu20
    R Version       : 4.0.3

Verified that plot outputs on ubuntu20 RStudio desktop show in high resolution. Also verified that plots outputs on RStudio Workbench show in high resolution (viewed through chromium browser on mac and windows).

**Comment by jmcphers** [comment]
@valerie-rstudio I tried on Windows and got a high resolution plot. I think this may be related to the scale factor in Windows -- it likely needs to be set at >= 200% for Windows to report a device pixel ratio of 2 (which is what in turn enables high resolution plotting). What's your scale factor set to?

**Comment by valerie-rstudio** [comment]
@jmcphers My Windows resolution is at 2560 x 1600, and I changed my scale to 200%. I'm still seeing a noticeable difference between the Mac and Windows plot displays. While the Mac plot displays seem smooth:
<img width="190" alt="MacPlotRes" src="https://user-images.githubusercontent.com/75644399/122788585-17ae2b00-d27c-11eb-8503-eb1f400213af.png">
I can see jagged pixelated edges on some of the Windows plot characters:
<img width="190" alt="WindowsPlotRes" src="https://user-images.githubusercontent.com/75644399/122788672-2e548200-d27c-11eb-8e25-c6a66d81600d.png">
I have tried multiple different combinations of resolution and scaling for Windows, but the characters in plots still do not appear smooth.

**Comment by jmcphers** [comment]
@valerie-rstudio Thank you for the screenshots! I think what you're seeing is a problem with the font rather than an overall resolution issue (the fonts built into Windows don't all have high resolution glyphs). Are plotted lines and curves appropriately smooth?

**Comment by valerie-rstudio** [comment]
@jmcphers Unfortunately it looks like the plotted lines & curves are also not smoothed for Windows (still on resolution of 2560 x 1600, and scale of 200%)

### Windows:
<img width="782" alt="WindowsPlotCurve" src="https://user-images.githubusercontent.com/75644399/122965386-c0788b00-d34d-11eb-8fbe-f4ee424ea779.png">

### Mac:
<img width="782" alt="MacPlotCurve" src="https://user-images.githubusercontent.com/75644399/122965421-ca01f300-d34d-11eb-90df-bbcc87a2061f.png">

### Code to generate plot:
```
set.seed(1919)
x1 <- rnorm(1000)
y1 <- x1 + rnorm(1000)
plot(density(x1))
```

**Comment by jmcphers** [comment]
>  (still on resolution of 2560 x 1600, and scale of 200%)

Maybe that's the difference, I'm working with a 4K display (2160 instead of 1600 pixels of vertical resolution). Ultimately Windows has to label the display to support a device pixel ratio of 2 for the high DPI plots to kick in. You can check that by right clicking anywhere in the IDE -> choose Inspect -> go to the Console tab -> execute the code `window.devicePixelRatio`. What does that give you?

**Comment by valerie-rstudio** [comment]
@jmcphers I figured out what was going wrong on my end. I had originally picked the highest available resolution and scaling in my Windows VM, assuming it would be the equivalent of my native display on my Mac. After seeing your comment, I dug into the VM configurations, and it turns out I had the graphics set to "Scaled" rather than "Best for Retina display". So at a resolution of 3584 x 2240 and scaling set to 350%, it looks like the high DPI plots finally kicked in. Closing this as fixed.
